src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala

```diff
@@ -891,52 +891,113 @@ abstract class BTypes {
   // otherwise the prefixes don't match and we get type mismatch errors.
   // Since we already do this dance (val bTypes: GenBCode.this.bTypes.type = GenBCode.this.bTypes)
   // for BTypes, it's easier to add those nested classes to BTypes.
 
-  object Lazy {
-    def apply[T <: AnyRef](t: => T): Lazy[T] = new Lazy[T](() => t)
+  abstract sealed class Lazy[T] {
+    /** get the result of the lazy value, calculation the result and performing the additional actions if the value
+      * is not already known.
+      * @return the
+      */
+    def force: T
+
+    /** add an accumulating action, which is performed when the result is forced.
+      * If the result is already known at the time of this call then perform the action immediately
+      *
+      * @param f the action
+      */
+    def onForce(f: T => Unit): Unit
   }
+  object Lazy {
 
-  /**
-   * A lazy value that synchronizes on the `frontendLock`, and supports accumulating actions
-   * to be executed when it's forced.
-   */
-  final class Lazy[T <: AnyRef](t: () => T) {
-    @volatile private var value: T = _
+    /**
+      * create a Lazy, whose calculation is performed with `frontendLock`
+      *
+      * @param t    the calculation
+      * @return a Lazy with which is guaranteed to perform if calculation with `frontendLock` locked
+      */
+    def withLock[T <: AnyRef](t: => T): Lazy[T] = new LazyWithLock[T](() => t)
+
+    /**
+      * create a Lazy where the result is pre-determined, typically a constant, e.g. Nil None etc
+      *
+      * @param value the final value of the Lazy
+      * @return a new Lazy, pre-initialised with the specified value
+      */
+    def eager[T <: AnyRef](value: T): Lazy[T] = new Eager[T](value)
 
-    private var initFunction = {
-      val tt = t // prevent allocating a field for t
-      () => { value = tt() }
+    /**
+      * create a Lazy, whose calculation is performed on demand
+      *
+      * @param t    the calculation
+      * @return a Lazy with which is guaranteed to perform if calculation with `frontendLock` locked
+      */
+    def withoutLock[T <: AnyRef](t: => T): Lazy[T] = new LazyWithoutLock[T](() => t)
+
+    private final class Eager[T](val force: T) extends Lazy[T] {
+      def onForce(f: T => Unit): Unit = f(force)
+
+      override def toString = force.toString
     }
 
-    override def toString = if (value == null) "<?>" else value.toString
+    private abstract class AbstractLazy[T <: AnyRef](t: () => T) extends Lazy[T] {
+      // value need be volatile to ensure that init doesn't expose incomplete results
+      // due to JVM inlining of init
+      @volatile protected var value: T = _
 
-    def onForce(f: T => Unit): Unit = {
-      if (value != null) f(value)
-      else frontendSynch {
+      override def toString = if (value == null) "<?>" else value.toString
+
+      private var postForce: List[T => Unit] = Nil
```

Comment:
The previous implementation had the advantage of not requiring a field for those actions - could we still use that approach now?

Reply:
I don't think that was a true saving. You had a inner class that wrapped the action via a dynamic class and it also extended the lock duration

If we are really concerned with these field counts we have save much more by making these top level classes